app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+        DocumentsContract.Root.COLUMN_SUMMARY,
+        DocumentsContract.Root.COLUMN_DOCUMENT_ID,
+        DocumentsContract.Root.COLUMN_AVAILABLE_BYTES
+    )
+    private val DEFAULT_DOCUMENT_PROJECTION : Array<String> = arrayOf(
+        DocumentsContract.Document.COLUMN_DOCUMENT_ID,
+        DocumentsContract.Document.COLUMN_MIME_TYPE,
+        DocumentsContract.Document.COLUMN_DISPLAY_NAME,
+        DocumentsContract.Document.COLUMN_LAST_MODIFIED,
+        DocumentsContract.Document.COLUMN_FLAGS,
+        DocumentsContract.Document.COLUMN_SIZE
+    )
+
+    override fun onCreate() : Boolean {
+        return true
+    }
+
+    override fun queryRoots(projection : Array<out String>?) : Cursor {
+        val result = MatrixCursor(projection ?: DEFAULT_ROOT_PROJECTION)
+        val applicationTitle = SkylineApplication.applicationContext().applicationInfo.loadLabel(SkylineApplication.applicationContext().packageManager).toString()
+
+        result.newRow().apply {
+            add(DocumentsContract.Root.COLUMN_ROOT_ID, getDocIdForFile(baseDir))
+            add(DocumentsContract.Root.COLUMN_SUMMARY, null)
+            add(
+                DocumentsContract.Root.COLUMN_FLAGS,
+                DocumentsContract.Root.FLAG_SUPPORTS_CREATE or
+                        DocumentsContract.Root.FLAG_SUPPORTS_SEARCH
+            )
+            add(DocumentsContract.Root.COLUMN_TITLE, applicationTitle)
+            add(DocumentsContract.Root.COLUMN_DOCUMENT_ID, getDocIdForFile(baseDir))
+            add(DocumentsContract.Root.COLUMN_MIME_TYPES, "*/*")
+            add(DocumentsContract.Root.COLUMN_AVAILABLE_BYTES, baseDir.freeSpace)
+            add(DocumentsContract.Root.COLUMN_ICON, null)
+        }
+        return result
+    }
+
+    private fun getDocIdForFile(file : File) : Any? {
+        return file.absolutePath
+    }
+
+    override fun queryDocument(documentId : String?, projection : Array<out String>?) : Cursor {
+        var result = MatrixCursor(projection ?: DEFAULT_DOCUMENT_PROJECTION)
+        result = includeFile(result, documentId, null)
+        return result
+    }
+
+    override fun isChildDocument(parentDocumentId : String?, documentId : String?) : Boolean {
+        return documentId?.startsWith(parentDocumentId!!) ?: false
+    }
+
+    override fun createDocument(parentDocumentId : String?, mimeType : String?, displayName : String) : String? {
+        var newFile = File(parentDocumentId, displayName)
+        var noConflictId = 2
+        while (newFile.exists()) {
+            newFile = File(parentDocumentId, displayName + " (" + noConflictId++ + ")")
+        }
+        try {
+            val succeeded : Boolean = if (DocumentsContract.Document.MIME_TYPE_DIR == mimeType) {
+                newFile.mkdir()
+            } else {
+                newFile.createNewFile()
+            }
+            if (!succeeded) {
+                throw FileNotFoundException("Failed to create document with id " + newFile.path)
+            }
```

Comment:
These are all control flow statements with a single-line body, we omit the `{}` braces in this case.